File: src/retar/role-sets.ts
/**
 * 役職集合 helper. systemRoles を ground truth に派生。
 *
 * 役職追加時は src/types/index.ts の systemRoles に entry を埋めるだけで
 * これらの helper は自動追従する (paparazzi 追加で Liar 欠落バグを踏んだ反省)。
 *
 * - `*In(setup)`: setup に含まれる役職 (count > 0) だけにフィルタ。
 *   探索木に登場しない役職を枝刈りでき、 LiarRoles 等の用途に。
 * - `allKnownRoles()`: systemRoles に登録されている全役職 (setup-free)。
 *   配役生成や bit 配置など、 全候補が必要なケースに。
 */

import { systemRoles, type SystemRole, type RoleTrait, type EnumSpecies } from '../types/index.ts'

/**
 * 指定 role が trait (kind + sub) を持つかチェック.
 * 役職追加で systemRoles の traits が増えれば自動追従する.
 */
export function hasTrait(role: SystemRole, kind: RoleTrait['kind'], sub: RoleTrait['sub']): boolean {
  return systemRoles.get(role)!.traits.some(t => t.kind === kind && t.sub === sub)
}

/** setup に含まれる「trait predicate を満たす役職」のリスト. */
export function rolesWithTraitIn(
  setup: Map<SystemRole, number>,
  kind: RoleTrait['kind'],
  sub: RoleTrait['sub'],
): SystemRole[] {
  return allRolesIn(setup).filter(role => hasTrait(role, kind, sub))
}

/** setup に含まれる「seerResult が指定種別の役職」のリスト. */
export function rolesBySeerResultIn(
  setup: Map<SystemRole, number>,
  result: EnumSpecies,
): SystemRole[] {
  return allRolesIn(setup).filter(role => systemRoles.get(role)!.seerResult === result)
}

/**
 * trait predicate を満たす役職を systemRoles から全て返す (setup 非依存).
 * rolesWithTraitIn の setup フィルタ無し版. fixedPositions の確定先など、
 * setup に存在しなくても systemRoles レベルで意味のある役職を取り出す場面で使う.
 */
export function rolesByTrait(kind: RoleTrait['kind'], sub: RoleTrait['sub']): SystemRole[] {
  return allKnownRoles().filter(r => hasTrait(r, kind, sub))
}

/**
 * trait predicate を満たす役職を systemRoles から 1 つだけ返す (setup 非依存).
 * 0 件 or 複数件で throw (= 設計上 1 件しか無い前提のロジック箇所で使う).
 * 例: passive:die-when-divined → werehamster.
 */
export function singleRoleByTrait(kind: RoleTrait['kind'], sub: RoleTrait['sub']): SystemRole {
  const matched = allKnownRoles().filter(r => hasTrait(r, kind, sub))
  if (matched.length !== 1) {
    throw new Error(`singleRoleByTrait(${kind}:${sub}) expected exactly 1 role, got ${matched.length}: ${matched.join(',')}`)
  }
  return matched[0]
}

/**
 * seerResult が指定種別の役職を systemRoles から全て返す (setup 非依存).
 * rolesBySeerResultIn の setup フィルタ無し版.
 */
export function rolesBySeerResult(result: EnumSpecies): SystemRole[] {
  return allKnownRoles().filter(r => systemRoles.get(r)!.seerResult === result)
}

/**
 * seerResult が指定種別の役職を systemRoles から 1 つだけ返す (setup 非依存).
 * 0 件 or 複数件で throw.
 * 例: seerResult='wolf' → werewolf.
 */
export function singleRoleBySeerResult(result: EnumSpecies): SystemRole {
  const matched = allKnownRoles().filter(r => systemRoles.get(r)!.seerResult === result)
  if (matched.length !== 1) {
    throw new Error(`singleRoleBySeerResult(${result}) expected exactly 1 role, got ${matched.length}: ${matched.join(',')}`)
  }
  return matched[0]
}

/** setup に含まれる「trait を持つ役職」の count 合計. */
export function countByTraitIn(
  setup: Map<SystemRole, number>,
  kind: RoleTrait['kind'],
  sub: RoleTrait['sub'],
): number {
  return rolesWithTraitIn(setup, kind, sub).reduce((sum, role) => sum + (setup.get(role) ?? 0), 0)
}

/** setup に含まれる「seerResult が指定種別の役職」の count 合計. */
export function countBySeerResultIn(
  setup: Map<SystemRole, number>,
  result: EnumSpecies,
): number {
  return rolesBySeerResultIn(setup, result).reduce((sum, role) => sum + (setup.get(role) ?? 0), 0)
}

/**
 * predicate を満たす役職を systemRoles から 1 つだけ返す (setup 非依存).
 * 0 件 or 複数件で throw. trait + faction 等の複合条件で 1 役職を取り出すケースに.
 */
export function singleRoleByPredicate(predicate: (role: SystemRole) => boolean): SystemRole {
  const matched = allKnownRoles().filter(predicate)
  if (matched.length !== 1) {
    throw new Error(`singleRoleByPredicate expected exactly 1 role, got ${matched.length}: ${matched.join(',')}`)
  }
  return matched[0]
}

/** systemRoles に登録されている全役職 (setup 非依存). 宣言順を保つ. */
export function allKnownRoles(): SystemRole[] {
  return Array.from(systemRoles.keys())
}

/** systemRoles に登録されている全 村陣営役職 (setup 非依存). 配役生成等で使う. */
export function allVillageRoles(): SystemRole[] {
  return allKnownRoles().filter(r => systemRoles.get(r)!.faction === 'village')
}

/** systemRoles に登録されている全 人外陣営役職 (= faction !== 'village'、 setup 非依存). */
export function allLiarRoles(): SystemRole[] {
  return allKnownRoles().filter(r => systemRoles.get(r)!.faction !== 'village')
}

/** setup に含まれる役職 (count > 0). */
export function allRolesIn(setup: Map<SystemRole, number>): SystemRole[] {
  const out: SystemRole[] = []
  for (const [role, count] of setup) {
    if (count > 0) out.push(role)
  }
  return out
}

/** setup に含まれる村陣営 (faction === 'village') 役職. */
export function villageRolesIn(setup: Map<SystemRole, number>): SystemRole[] {
  return allRolesIn(setup).filter(role => systemRoles.get(role)!.faction === 'village')
}

/**
 * setup に含まれる嘘つき (faction !== 'village') 役職.
 * 旧 planBuilder.ts の `LiarRoles` を setup フィルタ + systemRoles 派生にしたもの.
 */
export function liarRolesIn(setup: Map<SystemRole, number>): SystemRole[] {
  return allRolesIn(setup).filter(role => systemRoles.get(role)!.faction !== 'village')
}

/**
 * setup に含まれる人間種別 (seerResult === 'human') 役職.
 * 旧 retar/index.ts の `HumanRoles` を setup フィルタ + systemRoles 派生にしたもの.
 */
export function humanRolesIn(setup: Map<SystemRole, number>): SystemRole[] {
  return allRolesIn(setup).filter(role => systemRoles.get(role)!.seerResult === 'human')
}

/**
 * setup に含まれる「能力を持つ村陣営役職」(faction=village かつ traits を持つ).
 * 旧 planBuilder.ts の `rolesInTestPlanning` を setup フィルタ + systemRoles 派生にしたもの.
 * 例: seer / medium / bodyguard / mason / nekomata. villager (traits=[]) は除外.
 */
export function poweredVillageRolesIn(setup: Map<SystemRole, number>): SystemRole[] {
  return allRolesIn(setup).filter(role => {
    const r = systemRoles.get(role)!
    return r.faction === 'village' && r.traits.length > 0
  })
}

File: src/retar-rs/src/role_sets.rs
//! 役職集合 helper. systemRoles (Rust では SystemRole::ALL + faction()/seer_result()/traits())
//! を ground truth に派生する.
//!
//! 役職追加時は src/retar-rs/src/types.rs の SystemRole enum に variant を追加するだけで
//! これらの helper は自動追従する.
//!
//! TS 側 src/retar/role-sets.ts と関数名・semantics を 1:1 で揃える (sync-check 規約).

use crate::types::{EnumSpecies, Faction, RoleTrait, SystemRole};
use std::collections::BTreeMap;

/// systemRoles に登録されている全役職 (setup 非依存). 宣言順を保つ.
pub fn all_known_roles() -> Vec<SystemRole> {
    SystemRole::ALL.to_vec()
}

/// systemRoles に登録されている全 村陣営役職 (setup 非依存).
pub fn all_village_roles() -> Vec<SystemRole> {
    SystemRole::ALL
        .iter()
        .copied()
        .filter(|r| r.faction() == Faction::Village)
        .collect()
}

/// systemRoles に登録されている全 人外陣営役職 (= faction !== village, setup 非依存).
pub fn all_liar_roles() -> Vec<SystemRole> {
    SystemRole::ALL
        .iter()
        .copied()
        .filter(|r| r.faction() != Faction::Village)
        .collect()
}

/// setup に含まれる役職 (count > 0).
pub fn all_roles_in(setup: &BTreeMap<SystemRole, u32>) -> Vec<SystemRole> {
    setup
        .iter()
        .filter_map(|(&role, &count)| if count > 0 { Some(role) } else { None })
        .collect()
}

/// setup に含まれる村陣営役職.
pub fn village_roles_in(setup: &BTreeMap<SystemRole, u32>) -> Vec<SystemRole> {
    all_roles_in(setup)
        .into_iter()
        .filter(|r| r.faction() == Faction::Village)
        .collect()
}

/// setup に含まれる嘘つき役職 (faction != village).
/// 旧 plan_builder.rs の LIAR_ROLES を setup フィルタ + 派生にしたもの.
pub fn liar_roles_in(setup: &BTreeMap<SystemRole, u32>) -> Vec<SystemRole> {
    all_roles_in(setup)
        .into_iter()
        .filter(|r| r.faction() != Faction::Village)
        .collect()
}

/// setup に含まれる人間種別 (seerResult == Human) 役職.
pub fn human_roles_in(setup: &BTreeMap<SystemRole, u32>) -> Vec<SystemRole> {
    all_roles_in(setup)
        .into_iter()
        .filter(|r| r.seer_result() == EnumSpecies::Human)
        .collect()
}

/// setup に含まれる「通常 CO 経由で plan する村陣営能力者」(faction=village かつ非 passive
/// trait を持つ + contractor 除外).
/// 旧 plan_builder.rs の ROLES_IN_TEST_PLANNING を setup フィルタ + 派生にしたもの.
///
/// 非 passive trait = action / auto-info / knowledge / reactive / channel.
/// passive のみの trait (例: contractor の pair-required) は setup 制約であって
/// 役職テスト対象の能力ではないため planning 対象から除外する.
///
/// contractor は宇理炎の能力で auto-info trait を持つが、 CO ではなく公示 (announce) で
/// 配置が確定する「公示専用役職」なので、 通常の planning から除外する。
/// 公示で全席確定された場合のみ plan_builder の announce-fixed 動的追加経路で planning に乗る。
pub fn powered_village_roles_in(setup: &BTreeMap<SystemRole, u32>) -> Vec<SystemRole> {
    all_roles_in(setup)
        .into_iter()
        .filter(|&r| {
            r.faction() == Faction::Village
                && r != SystemRole::Contractor
                && r.traits().iter().any(|t| !t.is_passive())
        })
        .collect()
}

/// 指定 role が trait を持つかチェック.
pub fn has_trait(role: SystemRole, t: RoleTrait) -> bool {
    role.traits().iter().any(|&x| x == t)
}

/// setup に含まれる「trait を持つ役職」のリスト.
pub fn roles_with_trait_in(setup: &BTreeMap<SystemRole, u32>, t: RoleTrait) -> Vec<SystemRole> {
    all_roles_in(setup)
        .into_iter()
        .filter(|&r| has_trait(r, t))
        .collect()
}

/// setup に含まれる「seerResult が指定種別の役職」のリスト.
pub fn roles_by_seer_result_in(
    setup: &BTreeMap<SystemRole, u32>,
    result: EnumSpecies,
) -> Vec<SystemRole> {
    all_roles_in(setup)
        .into_iter()
        .filter(|r| r.seer_result() == result)
        .collect()
}

/// trait を持つ役職を systemRoles から全て返す (setup 非依存).
/// roles_with_trait_in の setup フィルタ無し版. fixed_positions の確定先など、
/// setup に存在しなくても systemRoles レベルで意味のある役職を取り出す場面で使う.
pub fn roles_by_trait(t: RoleTrait) -> Vec<SystemRole> {
    SystemRole::ALL
        .iter()
        .copied()
        .filter(|&r| has_trait(r, t))
        .collect()
}

/// trait を持つ役職を systemRoles から 1 つだけ返す (setup 非依存). 0 件 or 複数件で panic.
pub fn single_role_by_trait(t: RoleTrait) -> SystemRole {
    let matched: Vec<SystemRole> = SystemRole::ALL
        .iter()
        .copied()
        .filter(|&r| has_trait(r, t))
        .collect();
    assert_eq!(
        matched.len(),
        1,
        "single_role_by_trait({:?}) expected exactly 1 role, got {}: {:?}",
        t,
        matched.len(),
        matched
    );
    matched[0]
}

/// seerResult が指定種別の役職を systemRoles から全て返す (setup 非依存).
/// roles_by_seer_result_in の setup フィルタ無し版.
pub fn roles_by_seer_result(result: EnumSpecies) -> Vec<SystemRole> {
    SystemRole::ALL
        .iter()
        .copied()
        .filter(|&r| r.seer_result() == result)
        .collect()
}

/// seerResult が指定種別の役職を systemRoles から 1 つだけ返す (setup 非依存).
pub fn single_role_by_seer_result(result: EnumSpecies) -> SystemRole {
    let matched: Vec<SystemRole> = SystemRole::ALL
        .iter()
        .copied()
        .filter(|r| r.seer_result() == result)
        .collect();
    assert_eq!(
        matched.len(),
        1,
        "single_role_by_seer_result({:?}) expected exactly 1 role, got {}: {:?}",
        result,
        matched.len(),
        matched
    );
    matched[0]
}

/// predicate を満たす役職を systemRoles から 1 つだけ返す (setup 非依存).
pub fn single_role_by_predicate(predicate: impl Fn(SystemRole) -> bool) -> SystemRole {
    let matched: Vec<SystemRole> = SystemRole::ALL.iter().copied().filter(|&r| predicate(r)).collect();
    assert_eq!(
        matched.len(),
        1,
        "single_role_by_predicate expected exactly 1 role, got {}: {:?}",
        matched.len(),
        matched
    );
    matched[0]
}

/// setup に含まれる「trait を持つ役職」の count 合計.
pub fn count_by_trait_in(setup: &BTreeMap<SystemRole, u32>, t: RoleTrait) -> u32 {
    roles_with_trait_in(setup, t)
        .into_iter()
        .map(|r| setup.get(&r).copied().unwrap_or(0))
        .sum()
}

/// setup に含まれる「seerResult が指定種別の役職」の count 合計.
pub fn count_by_seer_result_in(setup: &BTreeMap<SystemRole, u32>, result: EnumSpecies) -> u32 {
    roles_by_seer_result_in(setup, result)
        .into_iter()
        .map(|r| setup.get(&r).copied().unwrap_or(0))
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeMap;

    fn make(entries: &[(SystemRole, u32)]) -> BTreeMap<SystemRole, u32> {
        entries.iter().copied().collect()
    }

    #[test]
    fn all_known_roles_contains_paparazzi() {
        let all = all_known_roles();
        assert!(all.contains(&SystemRole::Paparazzi));
        assert!(all.len() >= 12);
    }

    #[test]
    fn liar_roles_in_filters_by_setup() {
        let setup = make(&[(SystemRole::Villager, 3), (SystemRole::Werewolf, 1)]);
        let got = liar_roles_in(&setup);
        assert_eq!(got, vec![SystemRole::Werewolf]);
        assert!(!got.contains(&SystemRole::Paparazzi));
    }

    #[test]
    fn single_role_by_seer_result_wolf_is_werewolf() {
        assert_eq!(single_role_by_seer_result(EnumSpecies::Wolf), SystemRole::Werewolf);
    }

    #[test]
    fn single_role_by_trait_die_when_divined_is_werehamster() {
        assert_eq!(
            single_role_by_trait(RoleTrait::PassiveDieWhenDivined),
            SystemRole::Werehamster
        );
    }

    #[test]
    fn count_by_trait_in_follow_fox_death() {
        let setup = make(&[(SystemRole::Villager, 3), (SystemRole::Immoralist, 2)]);
        assert_eq!(count_by_trait_in(&setup, RoleTrait::ReactiveFollowFoxDeath), 2);
    }
}
